src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java

```diff
@@ -51,73 +52,78 @@ public UploadLettersTask(
     @SchedulerLock(name = TASK_NAME)
     @Scheduled(fixedDelayString = "${tasks.upload-letters.interval-ms}")
     public void run() {
-        if (!availabilityChecker.isFtpAvailable(now(ZoneId.of(EUROPE_LONDON)).toLocalTime())) {
-            logger.info("Not processing '{}' task due to FTP downtime window", TASK_NAME);
-            return;
-        }
-
         logger.info("Started '{}' task", TASK_NAME);
 
-        List<Letter> lettersToUpload = repo.findFirst3ByStatus(LetterStatus.Created);
-        int counter = 0;
-
-        if (!lettersToUpload.isEmpty()) {
-            counter = ftp.runWith(sftpClient -> uploadLetters(lettersToUpload, sftpClient));
+        if (!availabilityChecker.isFtpAvailable(now(ZoneId.of(EUROPE_LONDON)).toLocalTime())) {
+            logger.info("Not processing '{}' task due to FTP downtime window", TASK_NAME);
+        } else {
+            if (repo.countByStatus(LetterStatus.Created) > 0) {
+                int uploadCount = processLetters();
+                logger.info("Completed '{}' task. Uploaded {} letters", TASK_NAME, uploadCount);
+            } else {
+                logger.info("Completed '{}' task. No letters to upload.", TASK_NAME);
+            }
         }
-
-        logger.info("Completed '{}' task. Uploaded {} letters", TASK_NAME, counter);
     }
 
-    private int uploadLetters(List<Letter> lettersToUpload, SFTPClient sftpClient) {
-        return lettersToUpload
-            .stream()
-            .mapToInt(letter -> {
-                boolean isUploaded = uploadToFtp(letter, sftpClient);
-
-                if (isUploaded) {
-                    letter.setStatus(LetterStatus.Uploaded);
-                    letter.setSentToPrintAt(now());
+    private int processLetters() {
+        return ftp.runWith(client -> {
+            int uploadCount = 0;
+
+            for (int i = 0; i < BATCH_SIZE; i++) {
+                Optional<Letter> letter = repo.findFirstByStatusOrderByCreatedAtAsc(LetterStatus.Created);
+                if (letter.isPresent()) {
+                    boolean uploaded = processLetter(letter.get(), client);
+                    if (uploaded) {
+                        uploadCount++;
+                    }
                 } else {
-                    letter.setStatus(LetterStatus.Skipped);
+                    break;
                 }
+            }
```

Comment:
Bah, java9 here would be very handy with `.takeWhile` function

Reply:
soon ;)